include/aws/common/xml_parser.h

```diff
@@ -24,62 +23,51 @@ struct aws_xml_attribute {
 /**
  * Callback for when an xml node is encountered in the document. As a user you have a few options:
  *
- * 1. reject the document parsing at this point by returning false. This will immediately stop doc parsing.
+ * 1. fail the parse by returning AWS_OP_ERR (after an error has been raised). This will stop any further parsing.
  * 2. call aws_xml_node_traverse() on the node to descend into the node with a new callback and user_data.
  * 3. call aws_xml_node_as_body() to retrieve the contents of the node as text.
  *
+ * You MUST NOT call both aws_xml_node_traverse() and aws_xml_node_as_body() on the same node.
+ *
  * return true to continue the parsing operation.
  */
-typedef bool(
-    aws_xml_parser_on_node_encountered_fn)(struct aws_xml_parser *parser, struct aws_xml_node *node, void *user_data);
+typedef int(aws_xml_parser_on_node_encountered_fn)(struct aws_xml_node *node, void *user_data);
```

Comment:
So, if user wants to just stop parsing without error, they cannot do it now?

I guess we still have use case like I found the response I want and, now we can just stop parsing and it's not an error?

Reply:
Oh, I saw the description.

To me, I feel like the complexity is not too bad, just one more boolean, while it helps to improve performance? I just don't like the idea of having to do the extra parsing.

Reply:
Yes. Only errors can stop the parser now.

I found that in all cases but one, we were returning false because an error happened.

My 1st pass did have an extra `bool *stop_parsing` param that the callback could set, but I got sick of ignoring the param (`(void)stop_parsing;`) in every callback. Also, the `stop_parsing` code paths were not well tested. I felt better just removing the feature.

It was only used in the `aws_xml_get_top_level_tag()` functions. I changed it so its callback just immediately returns if its called again after finding the one thing it's looking for.

But in most cases we're gathering data from a few elements (e.g. credentials provider needs access key, secret, expiration, etc) and didn't use it